UGENE-7464. "Select manually" leads to the incorrectly selected region

evelinabis opened the pull request:
[UGENE-7464](https://ugene.dev/tracker/browse/UGENE-7464). Tests passed.

Commit:
UGENE-7464. "Select manually" leads to the incorrectly selected region.

Fixed.
Added GUI test.
Description of GUI test utility method has been updated.

Commit:
UGENE-7464. "Select manually" leads to the incorrectly selected region.

Fixed.
Added GUI test.
Description of GUI test utility method has been updated.

evelinabis committed:
Merge branch 'UGENE-7464' of https://github.com/ugeneunipro/ugene into UGENE-7464

yalgaer commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -274,7 +274,10 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_selectManually() {
     int endRegionValue = sbEndRegion->value();
-    sequenceSelection->setSelectedRegions({ U2Region(startRegionValue, endRegionValue - startRegionValue) });
+    if (startRegionValue > 0) {
+        sequenceSelection->setSelectedRegions({U2Region(static_cast<qint64>(startRegionValue) - 1,
```
nit: You do not need these ugly cast when you move from int32 to int64

evelinabis replied:
Without cast there is the warning C26451 "Arithmetic overflow: Using operator '-' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '-' to avoid overflow (io.2)".

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/pcr_primer_design_for_dna_assembly/GTTestsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -298,6 +298,45 @@ GUI_TEST_CLASS_DEFINITION(test_0007) {
+    GTUtilsProject::openFileExpectSequence(os, testDir + "_common_data/pcr_primer_design/gfp.fa", "gfp");
+    // Open the PCR Primer Design tab.
+    GTUtilsPcrPrimerDesign::openTab(os);
+    QWidget *sequence = GTUtilsSequenceView::getActiveSequenceViewWindow(os);
```
nit: use "auto sequenceWidget" instead of "Widget sequence"
This way the rest of the code will be more readable.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/pcr_primer_design_for_dna_assembly/GTTestsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -298,6 +298,45 @@ GUI_TEST_CLASS_DEFINITION(test_0007) {
+    // Open the PCR Primer Design tab.
+    GTUtilsPcrPrimerDesign::openTab(os);
+    QWidget *sequence = GTUtilsSequenceView::getActiveSequenceViewWindow(os);
+    QToolButton *selectManually = GTWidget::findToolButton(os, "tbLeftAreaSelectManually", sequence);
```
nit: use auto - you already have a type info (ToolButton) in the line.

yalgaer commented on src/plugins/GUITestBase/src/tests/common_scenarios/pcr_primer_design_for_dna_assembly/GTTestsPcrPrimerDesignForDnaAssembly.cpp:
```diff
@@ -298,6 +298,45 @@ GUI_TEST_CLASS_DEFINITION(test_0007) {
+                  QString("Selected region: expected 355-426, current: %1").arg(region.toString()))
+    GTUtilsSequenceView::selectSequenceRegion(os, 1, 1);
+    GTWidget::click(os, selectManually);
+    GTUtilsPcrPrimerDesign::setSearchArea(os, {{717, 717}, true}, GTUtilsPcrPrimerDesign::AreaType::Right);
```
nit: Does this line check anything? At least it does not look like a check.

evelinabis replied:
https://github.com/ugeneunipro/ugene/blob/5e746404528ceb3ce7391d814e7b91a96294ddae/src/plugins/GUITestBase/src/tests/common_scenarios/pcr_primer_design_for_dna_assembly/GTTestsPcrPrimerDesignForDnaAssembly.cpp#L319-L324
By my own logic, with these lines I check absence of a crash, although in general there have never been any problems in this place. If you think this is inappropriate, then I do not mind removing these lines, because it is logical.

yalgaer reviewed (commented):
No blockers, code style suggestions only.
You do not need my approve to merge it to the non-master/release branch.

evelinabis committed:
UGENE-7464. "Select manually" leads to the incorrectly selected region.

Code style changes by review.

evelinabis commented:
> No blockers, code style suggestions only. You do not need my approve to merge it to the non-master/release branch.

A review request to you as the codeowner is sent automatically.

DmitriiSukhomlinov commented on src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp:
```diff
@@ -274,7 +274,9 @@ void PCRPrimerDesignForDNAAssemblyOPWidget::sl_selectManually() {
     int startRegionValue = sbStartRegion->value();
     int endRegionValue = sbEndRegion->value();
-    sequenceSelection->setSelectedRegions({ U2Region(startRegionValue, endRegionValue - startRegionValue) });
+    if (startRegionValue > 0) {
```
In what case `startRegionValue` less than 0?

evelinabis replied:
With the current changes, the invalid region is not displayed in the Sequence View, before the changes, the region went beyond sequence boundary.

DmitriiSukhomlinov reviewed (changes requested):
Scenario:
1. Open gfp.fa
2. Open the "PCR Primer Design" tab
3. Click "Select manually" (any of them)
4. Click "Ctrl + A" and choose 1 as "Min" and 1 as "Max"
Expected: The corresponding area on the tab has `1 - 1`
Current: The corresponding area on the tab has `0 - 1`

There is the problem with the selection 1 nucleotide length

evelinabis commented:
> Scenario:
> 
> 1. Open gfp.fa
> 2. Open the "PCR Primer Design" tab
> 3. Click "Select manually" (any of them)
> 4. Click "Ctrl + A" and choose 1 as "Min" and 1 as "Max"
>    Expected: The corresponding area on the tab has `1 - 1`
>    Current: The corresponding area on the tab has `0 - 1`
> 
> There is the problem with the selection 1 nucleotide length

I didn't spawn this bug. You can create a new issue for it if you like, since it is not associated with [issue 7464](https://ugene.dev/tracker/browse/UGENE-7464).

You wrote that an area [must be](https://github.com/ugeneunipro/ugene/blame/45862ffbdddc210dea6d7fccdbef61bf465ed80d/src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp#L303-L305) at least two in length. Should I completely change your behavior or just for specific case 1-1?

evelinabis commented:
> > Scenario:
> > 
> > 1. Open gfp.fa
> > 2. Open the "PCR Primer Design" tab
> > 3. Click "Select manually" (any of them)
> > 4. Click "Ctrl + A" and choose 1 as "Min" and 1 as "Max"
> >    Expected: The corresponding area on the tab has `1 - 1`
> >    Current: The corresponding area on the tab has `0 - 1`
> > 
> > There is the problem with the selection 1 nucleotide length
> 
> I didn't spawn this bug. You can create a new issue for it if you like, since it is not associated with [issue 7464](https://ugene.dev/tracker/browse/UGENE-7464).
> 
> You wrote that an area [must be](https://github.com/ugeneunipro/ugene/blame/45862ffbdddc210dea6d7fccdbef61bf465ed80d/src/plugins/pcr_primer_design_for_dna_assembly/src/PCRPrimerDesignForDNAAssemblyOPWidget.cpp#L303-L305) at least two in length. Should I completely change your behavior or just for specific case 1-1?

@DmitriiSukhomlinov, please answer the question.

yalgaer commented:
Closing as inactive. No code updates in almost 2 months

evelinabis committed:
UGENE-7464. "Select manually" leads to the incorrectly selected region.

The area spinboxes are separate from the rest of tab spinboxes.
Changed area spinbox range policy.
Added a helper function to test the tab.
Updated GUI tests for the tab.
Added a GUI test for the tab.

evelinabis committed:
Merge remote-tracking branch 'origin/pcr_primer_design_for_dna_assembly' into UGENE-7464

evelinabis commented:
Waiting for GUI tests
